### PR TITLE
Kim 159813379 rate engine to edi generation

### PR DESCRIPTION
**Pull request by kimallen**
## Description

This allows a shipment (with an accepted ShipmentOffer) to pass through the rate engine, return a value for cost, and pass the values of cost and shipment as a struct to the edi generator to return a an edi.  It also splits out the pack/unpack costs into separate line items in the generator (instead of combining them in one line item).

## Reviewer Notes

The current test for the generator passes an empty cost and only checks that a string is returned and not an error.  Are there other tests that should be written?

## Setup
1. make tools_build
2.  Run make db_dev_reset && make db_dev_migrate && go run ./cmd/generate_test_data/main.go -scenario=7
3.  Find an ACCEPTED shipment, locate the shipment_offer connected to it, set accepted=true on that record.
4. Find the move_id on the shipment and use that to run the generate_shipment_edi program.

To run the command:
```golang
go run cmd/generate_shipment_edi/main.go -moveID <UUID>
```
It should return a mess of an EDI string

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/159813379) for this change

### CODE REVIEW

**Comment by chrisgilmerproj** [comment]
I tried to run this and it didn't work for me:

```
go run cmd/generate_shipment_edi/main.go -moveID AE8F98AF-3FC6-4C3E-AAAB-161A104B45C0
[POP] 2018/09/12 14:20:43 info - Loading config file from /Users/cgilmer/Projects/go/src/github.com/transcom/mymove/config/database.yml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x15671e3]

goroutine 1 [running]:
main.HandleRunRateEngineOnShipment(0xc44100f92636c42e, 0xed0c9775cf8c0189, 0xc4202f1420, 0x0, 0x7b4d6a15efa4e227, 0x8b59716e4fcc77b8, 0xc420318c00, 0xc4205c0380, 0xc4205c0360, 0x345e9208, ...)
        /Users/cgilmer/Projects/go/src/github.com/transcom/mymove/cmd/generate_shipment_edi/main.go:77 +0xb3
main.main()
        /Users/cgilmer/Projects/go/src/github.com/transcom/mymove/cmd/generate_shipment_edi/main.go:58 +0x57a
exit status 2
```

To reproduce:
1. Run `make db_dev_reset && make db_dev_migrate && go run ./cmd/generate_test_data/main.go -scenario=7`
2. Find an ACCEPTED shipment, locate the shipment_offer connected to it, set `accepted=True` on that record.
3. Find the move_id on the shipment and use that to run the `generate_shipment_edi` program.

Can you help me figure this out?

File: cmd/generate_shipment_edi/main.go [hunk 1]
```diff
@@ -28,14 +33,60 @@ func main() {
...
-	err = db.Where("move_id = $1", &moveID).All(&shipments)
+
+	err = db.Eager(
+		"Move.Orders",
+		"PickupAddress",
+		"DeliveryAddress",
+		"ServiceMember",
+	).Where("shipment_offers.accepted=true").
+		Where("move_id = $1", &moveID).
+		Join("shipment_offers", "shipment_offers.shipment_id = shipments.id").
```
Comment (chrisgilmerproj) [inline cmd/generate_shipment_edi/main.go]: I'm not sure that it matters but I feel like I always do the JOIN comments before WHERE comments in my sql creation.  I bet Pop does the right thing here but it stuck out to me.

File: cmd/generate_shipment_edi/main.go [hunk 2]
```diff
@@ -28,14 +33,60 @@ func main() {
...
+
+// HandleRunRateEngineOnShipment runs the rate engine on a shipment and returns the shipment and cost
+func HandleRunRateEngineOnShipment(shipment models.Shipment, engine *rateengine.RateEngine) (ediinvoice.CostByShipment, error) {
+	// Apply rate engine to shipment
+	var shipmentCost ediinvoice.CostByShipment
+	cost, err := engine.ComputeShipment(unit.Pound(*shipment.WeightEstimate),
+		shipment.PickupAddress.PostalCode,
+		shipment.DeliveryAddress.PostalCode,
+		time.Time(*shipment.PickupDate),
+		0,  // We don't want any SIT charges
```
Comment (chrisgilmerproj) [inline cmd/generate_shipment_edi/main.go]: I'd prefer these to be named variables that you plug in here instead of values.  Might make it read better.

**Comment by kimallen** [comment]
I ran into problems with regenerating the db and it not applying the changes I made to MakeShipment (added destination address).  I had to run make tools_build.  Run that and then try again.  The other steps seem right (and I'm going to add those to the setup instructions- thx!

File: cmd/generate_shipment_edi/main.go (hunk 2 above)
Comment (kimallen) [reply cmd/generate_shipment_edi/main.go]: That totally makes sense- changing

File: cmd/generate_shipment_edi/main.go (hunk 1 above)
Comment (kimallen) [reply cmd/generate_shipment_edi/main.go]: I tried moving JOIN before both WHEREs, but it didn't like that.   I moved one of the WHEREs, to after the JOIN, though maybe that's even more weird?

**Comment by chrisgilmerproj** [comment]
I got it to work!

```
 go run ./cmd/generate_shipment_edi/main.go -moveID='715E6F9F-BEA3-411C-A9FA-1CAC29823686'
[POP] 2018/09/12 15:31:10 info - Loading config file from /Users/cgilmer/Projects/go/src/github.com/transcom/mymove/config/database.yml
ISA~00~0000000000~00~0000000000~ZZ~W28GPR-DPS     ~12~8004171844     ~180912~1531~U~00401~000000001~1~T~|
GS~SI~W28GPR-DPS~8004171844~20180912~1531~1~X~004010
ST~858~0001
BX~00~J~PP~ABCD00001~MCCG~~4
N9~DY~SC~~
N9~CN~ABCD00001-1~~
N9~PQ~ABBV2708~~
N9~OQ~ORDER84~ARMY~20180315
N1~SF~Spacemen~~
N3~123 Any Street~P.O. Box 12345
N4~Beverly Hills~CA~90210~US~~
N1~RG~LKNQ~27~LKNQ
N1~RH~MLNQ~27~MLNQ
FA1~DZ
FA2~TA~tac-placeholder
L10~108.200~B~L
HL~303~~SS
L0~1~1.000~FR~0.000~~~~~~~
L1~0~0.0000~RC~202592~~~~~~~~LHS
HL~303~~SS
L0~1~0.000~~108.200~B~~~~~~L
L1~0~65.7700~RC~80568~~~~~~~~105A
HL~304~~SS
L0~1~0.000~~108.200~B~~~~~~L
L1~0~65.7700~RC~8459~~~~~~~~105C
HL~303~~SS
L0~1~0.000~~108.200~B~~~~~~L
L1~0~4.0700~RC~9300~~~~~~~~135A
HL~304~~SS
L0~1~0.000~~108.200~B~~~~~~L
L1~0~4.0700~RC~9300~~~~~~~~135B
HL~303~~SS
L0~1~1.000~FR~0.000~~~~~~~
L1~0~0.0300~RC~22742~~~~~~~~16A
SE~33~0001
GE~1~1
IEA~1~000000001
```

File: cmd/generate_shipment_edi/main.go (hunk 1 above)
Comment (chrisgilmerproj) [reply cmd/generate_shipment_edi/main.go]: Yeah, I'd say put this back the way you had it.  Sorry for the run around!

Comment (reggieriser) [reply cmd/generate_shipment_edi/main.go]: I think the reason it didn't work is that some of these Pop methods can be called on a `pop.Connection` struct, some can be called on a `pop.Query` struct, and some can be called on both.  The `Join` method is one that seems to only be available for calling on a `Query`.  In your code, `Eager` is called on a `Connection` and returns a `Connection`.  `Where` is called on a `Connection` and returns a `Query`.  So `Join` works fine coming after the `Where`, but if you put it before the `Where` (and after the `Eager`), you're trying to call it on a `Connection` and there's no `Join` that works on a `Connection`.

There's a `Q` method that you can call on a `Connection` to get an empty `Query` back that could help, but I don't know if you can do that in the middle of a big chain like this.

`Eager` does seem to have a version that is called on a `Query` and returns a `Query`, so you could do `db.Q().Eager(...).Join(...).Where(...).Where(...).All(&shipments)` I think.

I'll need to dig into Pop deeper to figure out what the logic is behind the different receiver types and return types.

Comment (chrisgilmerproj) [reply cmd/generate_shipment_edi/main.go]: This is a nice explanation.  Thanks!

Comment (kimallen) [reply cmd/generate_shipment_edi/main.go]: Aah, ok.  I wondered what the Q() was in other queries.  Thanks for the explanation

**Pull request merged**